include/net/nrf_cloud_codec.h

```diff
@@ -20,37 +27,535 @@ extern "C" {
  * @{
  */
 
+/** @brief Object types for encoding and decoding */
+enum nrf_cloud_obj_type {
+	NRF_CLOUD_OBJ_TYPE__UNDEFINED,
+
+	NRF_CLOUD_OBJ_TYPE_JSON,
+	NRF_CLOUD_OBJ_TYPE_CBOR,
+
+	NRF_CLOUD_OBJ_TYPE__LAST,
+};
+
+/** @brief Encoded data sources */
+enum nrf_cloud_enc_src {
+	/** Encoded data is not set */
+	NRF_CLOUD_ENC_SRC_NONE,
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
+};
+
+/** @brief Object used for building nRF Cloud messages. */
+struct nrf_cloud_obj {
+
+	enum nrf_cloud_obj_type type;
+	union {
+		cJSON *json;
+		// TODO
+		void *cbor; //struct zcbor_state_t cbor;
+	};
+
+	/** Source of encoded data  */
+	enum nrf_cloud_enc_src enc_src;
+	/** Encoded data */
+	struct nrf_cloud_data encoded_data;
+};
+
+/** @brief Define an nRF Cloud JSON object.
+ *
+ * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_JSON.
+ *
+ * @param _name	Name of the object.
+ */
+#define NRF_CLOUD_OBJ_JSON_DEFINE(_name) \
+	struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_JSON, .json = NULL, \
+				       .enc_src = NRF_CLOUD_ENC_SRC_NONE, \
+				       .encoded_data = { .ptr = NULL, .len = 0 } }
```

Comment:
I'm not 100% sure how I feel about this being a macro

On the one hand, it allows us to conveniently configure defaults, especially if any of them need to start out nonzero

On the other hand, I feel like user code would be more natural just straight up declaring `struct nrf_cloud_obj`

Overall I think I lean towards keeping the macros, but it still seems worthwhile to bring it up for discussion

Reply:
the macros were made to simplify struct initialization, so the user does not have to make sure they set each value correctly.
the user is free to declare and init the struct however they want.